# **ECS7020P Mini-Project**


### **Advanced component**

### **Author:** Sayali Vijay Ghodke
### **Student ID:** 230692620

## **Problem Formulation**

In this problem statement, we predict whether the particular dish is made at home or at a restaurant based on the ingredients of the dish. In this, we will be considering the ingredients as the input and classify the dish as **"Homemade"** or **"Restaurant"**.

As we know, there are certain ingredients that are usually not available at Home but always available in a Restaurant, so we have considered this problem statement in order to determine whether the dish is made with the ingredients available at home or the dish is made at a restaurant.

This will be helpfull in a real life scenario where we can detemine whether a dish can be made at home or not.


## Machine Learning Pipline:

### 1. Input Format:

We will be considering the attributes dataset for this problem statement. As we need to classify that the dish is **"Homemade"** or **"Restaurant_Made"**, we will need to focus on two columns **"Ingredients"**  and **"Home_or_restaurant"**. In the CSV file **"MLEndYD_image_attributes_benchmark.csv"** we have various columns such as **filename** (which is nothing but the names of the images), **Diet** (this column is a drop down option which has three options, Vegetarian, Non-vegetarian,
or Vegan), **Cuisine_org** (which contains the region of origin of the dish), Cuisine (this columns contains the name of the country in which the corresponding dish is a speciality), **Dish_name** (which contains names of the dishes that corresponds to the image), **Home_or_restaurant** (which states whether the corresponding dish is made at home or any restaurant, who's name is specified), **Ingredients** (it contains the basic ingredients that are used to make that specific dish), **Healthiness_rating** (this column is also drop down option which has, Very healthy, Healthy, Neutral, Unhealthy, Very unhealthy, which will decribe the healthyness of the dish)**Healthiness_rating_int** (this column contains the integer value for categorical values in Healthiness_rating column), **Likeness** (this is also a drop down column which will tell us about the likelihood of that dish based on a random person and it has, Strongly like, Like, Neutral, Dislike, Strongly Dislike), **Likeness_int** (this column contains the integer value for categorical values in Likeness column) and **Benchmark_A** (in this column we have determined whether the specific image is taken for training or testing).

We have now described the attribute dataset that we will have to consider in order to design the prediction model.

### 2. Preprocessing:

Let us now load the dataset:
        In order to load the dataset, we will make use of **google colab** and **google drive**. We will first make sure that we can access our google drive through google colab. To do so, we first created a seperate folder in our drive, and in that created another folder named 'Data', again created a folder inside 'Data', named 'MLEnd', and then we install the library 'mlend' as below.

In [1]:
!pip install mlend

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.6/1.6 MB 7.3 MB/s eta 0:00:00


Now, we will load the dataset. In order to do so, we will use different libraries and modules from **python**, and then mount our **google drive**.

In [2]:
from google.colab import drive

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import spkit as sp

from skimage import exposure
from skimage.color import rgb2hsv, rgb2gray
import skimage as ski

import mlend
from mlend import download_yummy, yummy_load

import os, sys, re, pickle, glob
import urllib.request
import zipfile

import IPython.display as ipd
from tqdm import tqdm
import librosa

drive.mount('/content/drive')

Mounted at /content/drive


Now, we will be loading the dataset:

In [3]:
subset = {}

datadir = download_yummy(save_to = '/content/drive/MyDrive/Data/MLEnd', subset = subset,verbose=1,overwrite=False)

100%|▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓|3250\3250|003250.jpg
Done!


Here, we are using **download_yummy** function to download the **MLEnd Yummy Dataset**. We have also saved it in a different directory.

Our attributes of the dataset are text based, which is why we will be creating a CSV file to store those attributes. We will not be storing the Photo attribute in this file as it is complex to be stored in such a file. So, we store the images in a different folder, but the filename of each image is still stored in the CSV file.

In [4]:
MLENDYD_df = pd.read_csv('/content/drive/MyDrive/Data/MLEnd/yummy/MLEndYD_image_attributes_benchmark.csv').set_index('filename')
MLENDYD_df

Diet Cuisine_org   Cuisine  \
filename                                           
000001.jpg  non_vegetarian    japanese  japanese   
000002.jpg  non_vegetarian     english   english   
000003.jpg  non_vegetarian     chinese   chinese   
000004.jpg      vegetarian      indian    indian   
000005.jpg  non_vegetarian      indian    indian   
...                    ...         ...       ...   
003246.jpg      vegetarian      indian    indian   
003247.jpg      vegetarian      indian    indian   
003248.jpg      vegetarian      indian    indian   
003249.jpg           vegan      indian    indian   
003250.jpg  non_vegetarian    american  american   

                                   Dish_name     Home_or_restaurant  \
filename                                                              
000001.jpg                chicken_katsu_rice          marugame_udon   
000002.jpg                 english_breakfast                   home   
000003.jpg                     spicy_chicken  jinli_flagship_branch   
000004.jpg                       gulab_jamun                   home   
000005.jpg                    chicken_masala                   home   
...                                      ...                    ...   
003246.jpg                        zeera_rice                   home   
003247.jpg                    paneer_and_dal                   home   
003248.jpg                            samosa                   home   
003249.jpg                        fruit_milk                   home   
003250.jpg  beef_burger_with_onion_and_salad                   home   

                                                  Ingredients  \
filename                                                        
000001.jpg              rice,chicken_breast,spicy_curry_sauce   
000002.jpg  eggs,bacon,hash_brown,tomato,bread,tomato,bake...   
000003.jpg  chili,chicken,peanuts,sihuan_peppercorns,green...   
000004.jpg      sugar,water,khoya,milk,salt,oil,cardamon,ghee   
000005.jpg  chicken,lemon,turmeric,garam_masala,coriander_...   
...                                                       ...   
003246.jpg  1_cup_basmati_rice,2_cups_water,2_tablespoons_...   
003247.jpg  fried_cottage_cheese,ghee,lentils,milk,wheat_f...   
003248.jpg  potato,onion,peanut,salt,turmeric_powder,red_c...   
003249.jpg                             kiwi,banana,apple,milk   
003250.jpg   beef_patty,bread_roll,cherry_tomato,_onion,chive   

           Healthiness_rating  Healthiness_rating_int       Likeness  \
filename                                                               
000001.jpg            neutral                     3.0           like   
000002.jpg          unhealthy                     2.0           like   
000003.jpg            neutral                     3.0  strongly_like   
000004.jpg          unhealthy                     2.0  strongly_like   
000005.jpg            healthy                     4.0  strongly_like   
...                       ...                     ...            ...   
003246.jpg            healthy                     4.0  strongly_like   
003247.jpg            healthy                     4.0  strongly_like   
003248.jpg     very_unhealthy                     1.0           like   
003249.jpg       very_healthy                     5.0  strongly_like   
003250.jpg            neutral                     3.0           like   

            Likeness_int Benchmark_A  
filename                              
000001.jpg           4.0       Train  
000002.jpg           4.0       Train  
000003.jpg           5.0       Train  
000004.jpg           5.0       Train  
000005.jpg           5.0       Train  
...                  ...         ...  
003246.jpg           5.0       Train  
003247.jpg           5.0        Test  
003248.jpg           4.0        Test  
003249.jpg           5.0       Train  
003250.jpg           4.0       Train  

[3250 rows x 11 columns]

Above are the contents of the CSV file. We can see that there are 3250 rows and 11 columns in our dataset, however, the cloumn **'filename'** acts as index and link to the image of the dish, which is why we have 12 columns. As we have already discussed, that we have cloumn **'Benchmark_A'**, in which we have determined whether the specific image is taken for training or testing. Using this column, let us divide the dataset into two different datasets, i.e. training and testing datasets. We will make the use of **yummy_load** function.

In the Basic Component of our project, we have used the in-built function **"yummy_small_load"** to load the dataset, by splitting it into train and test dataset and also mapping the values of **"Rice"** and **"Chips"** to **"0"** and **"1"** respectively. **But** here, we will have to customize the function in such a way that the **"home"** value of the **"Home_or_restaurant"** column is mapped to **"0"** and rest of all the values will be mapped to **"1"**. We also need the **"Ingredients"** column in order to design our problem statement. So, we will be customizing the **yummy_load** function as below:

In [5]:
from sklearn.model_selection import train_test_split as sklearn_train_test_split

def yummy_load(datadir_main='../MLEnd/yummy', train_test_split='Benchmark_A', verbose=1, encode_labels=True):

    repo_path = 'https://github.com/MLEndDatasets/Yummy'
    attributes_file_org = repo_path + '/raw/main/MLEndYD_image_attributes_benchmark.csv'

    img_files = glob.glob(os.path.join(datadir_main, 'MLEndYD_images', '*.jpg'))
    if verbose:
        print(f'Total {len(img_files)} found in {os.path.join(datadir_main, "MLEndYD_images")}/')

    attributes_file = os.path.join(datadir_main, 'MLEndYD_image_attributes_benchmark.csv')
    if os.path.isfile(attributes_file):
        D = pd.read_csv(attributes_file)
    else:
        D = pd.read_csv(attributes_file_org)

    img_files_map = {file.split('/')[-1]: file for file in img_files}

    Di = D.copy()

    X_train_filepath = []
    Y_train = []

    X_test_filepath = []
    Y_test = []

    if isinstance(train_test_split, str):
        if train_test_split in ['Benchmark_A']:
            D_train = Di[Di[train_test_split] == 'Train']
            D_test = Di[Di[train_test_split] == 'Test']
        elif train_test_split.lower() == 'random':
            Nt = int(Di.shape[0] * 0.7)
            idx = [True] * Nt + [False] * (Di.shape[0] - Nt)
            idx = np.array(idx)
            np.random.shuffle(idx)
            D_train = Di[idx]
            D_test = Di[~idx]
    elif isinstance(train_test_split, float) and 0 < train_test_split < 1:
        Nt = int(Di.shape[0] * train_test_split)
        idx = [True] * Nt + [False] * (Di.shape[0] - Nt)
        idx = np.array(idx)
        np.random.shuffle(idx)
        D_train = Di[idx]
        D_test = Di[~idx]
    else:
        raise ValueError("Invalid value for train_test_split. Should be str ['Benchmark_A', 'random'] or float (<1 and >0)")

    files_train = list(D_train['filename'])
    X_train_filepath = [img_files_map[file] for file in files_train]
    Y_train = np.array(D_train['Home_or_restaurant']).copy()

    files_test = list(D_test['filename'])
    X_test_filepath = [img_files_map[file] for file in files_test]
    Y_test = np.array(list(D_test['Home_or_restaurant'])).copy()

    TrainSet = {'X_paths': X_train_filepath, 'Y': Y_train, 'Ingredients': list(D_train['Ingredients'])}
    TestSet = {'X_paths': X_test_filepath, 'Y': Y_test, 'Ingredients': list(D_test['Ingredients'])}

    Y_train_enc = []
    Y_test_enc = []
    MAPs = {}

    if encode_labels:
        home_rest_map = {'home': 0}
        y1 = Y_train[:].copy()
        y2 = Y_test[:].copy()

        unique_labels = list(set(y1) | set(y2))
        unique_labels.sort()

        for k, v in enumerate(unique_labels):
            if v == 'home':
                home_rest_map[v] = 0
            else:
                home_rest_map[v] = 1

        Y_train_enc = np.array([home_rest_map[label] for label in y1])
        Y_test_enc = np.array([home_rest_map[label] for label in y2])

        MAPs = {'Home_or_restaurant': home_rest_map}

        TrainSet['Y_encoded'] = Y_train_enc
        TestSet['Y_encoded'] = Y_test_enc

    return TrainSet, TestSet, MAPs


In [6]:
TrainSet, TestSet, Map = yummy_load(datadir_main=datadir,train_test_split='Benchmark_A')

Total 3250 found in /content/drive/MyDrive/Data/MLEnd/yummy/MLEndYD_images/


Let us now describe how the train and test datasets are mapped with 0 and 1 instead of home and any other restaurant name.

In [7]:
TrainSet.keys()

dict_keys(['X_paths', 'Y', 'Ingredients', 'Y_encoded'])

In [8]:
TestSet.keys()

dict_keys(['X_paths', 'Y', 'Ingredients', 'Y_encoded'])

In [9]:
Map

{'Home_or_restaurant': {'home': 0,
  '36_street_food': 1,
  'afghan_grill': 1,
  'albura_kathisma_bistro': 1,
  'aldi': 1,
  'ambala': 1,
  'anjappar_restaurant': 1,
  'antalya_restaurant': 1,
  'asda': 1,
  'bamboo': 1,
  'bananatree': 1,
  "bill's": 1,
  'biryani_house': 1,
  'broasto': 1,
  'bui_vien': 1,
  'buns_&_berries': 1,
  'burgasum': 1,
  'burger_&_beyond': 1,
  'burger_and_chop': 1,
  'burgitzza_watford': 1,
  'cafe_de_nata': 1,
  'cake_box': 1,
  'canting': 1,
  'capital_restaurant': 1,
  'chengdufu': 1,
  'chicken_shop': 1,
  'chicken_world': 1,
  'chinese_supermarket': 1,
  'chongqing_old_street': 1,
  'co-op': 1,
  'coco_di_mama': 1,
  'costa': 1,
  'curve': 1,
  'dalchini': 1,
  'dandan_noodle_restaurant': 1,
  'dartmoor_lodge': 1,
  'dashi_chazuke': 1,
  'delamina_east': 1,
  'desi_lounge': 1,
  'dominos': 1,
  'domino’s': 1,
  'eat_tokyo': 1,
  'efes': 1,
  'efes_restaurant': 1,
  "emilia's_crafted_pasta": 1,
  'eminonu_borekcisi': 1,
  'emirates_restaurant': 1,
  'e

You can see in the above cell, the **"home"** value is mapped to **"0"** and rest of the values are mapped to **"1"**. We can now say that we have two classes. If we would not have done this, we would have had many different classes. You can see that there are also spelling mistakes which would have caused generation of extra unwanted classes.

Let us now make a variable which will store the list of training **"Ingredients"**, training **"Y_encoded"**, which is nothing but the binary structure of the **"Home_or_restaurant"** column.

In [10]:
TrainSet_df = pd.DataFrame(TrainSet)
TestSet_df = pd.DataFrame(TestSet)

X_train = TrainSet_df['Ingredients']
y_train = TrainSet_df['Y_encoded']

X_test = TestSet_df['Ingredients']
y_test = TestSet_df['Y_encoded']

Whenever we are taking care of any problem statement, we should always make sure that we do not have any type of null values. Let us take care of them first.

In [11]:
X_train = X_train.fillna('')
X_test = X_test.fillna('')

### 3. Feature Extraction OR Transformation Stage:

Let us now extract the features using **CountVectorizer** in order to convert the collection of text that is **"Ingredients"** into a matrix of token counts. This is done so that we can convert the categorical values of ingredients into numerical format.

In [12]:
from sklearn.feature_extraction.text import CountVectorizer

vectorizer = CountVectorizer()
X_train_vectorized = vectorizer.fit_transform(X_train)
X_test_vectorized = vectorizer.transform(X_test)

### 4. Trianing OR Modelling:

### Methodology:

First we will train this data with the help of **Multinomial Naive Bayes model** algorithm and check for its accuracy.

In [13]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

model = MultinomialNB()
model.fit(X_train_vectorized, y_train)

# Make predictions on the test set
y_pred = model.predict(X_test_vectorized)

# Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
conf_matrix = confusion_matrix(y_test, y_pred)
class_report = classification_report(y_test, y_pred)

print(f"Accuracy: {accuracy * 100:.2f}%")
print("Confusion Matrix:")
print(conf_matrix)
print("Classification Report:")
print(class_report)

Accuracy: 73.95%
Confusion Matrix:
[[588  65]
 [189 133]]
Classification Report:
              precision    recall  f1-score   support

           0       0.76      0.90      0.82       653
           1       0.67      0.41      0.51       322

    accuracy                           0.74       975
   macro avg       0.71      0.66      0.67       975
weighted avg       0.73      0.74      0.72       975



**73.95% accuracy!!!**

### 5. Prediction:

Quite good, but let us check what it predicts for a label by giving it some input.

In [14]:
test_input_index = 17  # You can change this to any index in the test set
input_test_data = X_test.iloc[[test_input_index]].values  # Get the input data for prediction
actual_label = y_test.iloc[test_input_index]  # Get the actual label

# Vectorize the input data
input_test_vectorized = vectorizer.transform(input_test_data)

# Make prediction
predicted_label = model.predict(input_test_vectorized)[0]

# Print ingredients and labels
print(f"Ingredients for Test Input {test_input_index}: {input_test_data[0]}")
print(f"Actual Label: {'Home' if actual_label == 0 else 'Restaurant'}")
print(f"Predicted Label: {'Home' if predicted_label == 0 else 'Restaurant'}")

Ingredients for Test Input 17: chicken,oil,bread_crumbs,flour,egg,honey_and_garlic_sauce
Actual Label: Restaurant
Predicted Label: Home


**OHH**

The actual and Predicted label for the second test data is not the same.

**BUT**

Have you noticed something?

We can see that in the set of input ingredients, we have **"honey_and_garlic_sauce"**, which is usually not present at home, and hence the Actual Label says that the dish is made at a Restaurant.

Now, let us try to predict the same using a different algorithm, say **Neural Networks**.

In [16]:
from sklearn.preprocessing import LabelEncoder
from sklearn.feature_extraction.text import CountVectorizer
from keras.models import Sequential
from keras.layers import Dense, Embedding, Flatten
from keras.utils import to_categorical

# Convert lists to DataFrames
TrainSet_df = pd.DataFrame(TrainSet)
TestSet_df = pd.DataFrame(TestSet)

# Handle missing values in the 'Ingredients' column for both training and test sets
TrainSet_df['Ingredients'] = TrainSet_df['Ingredients'].fillna('')
TestSet_df['Ingredients'] = TestSet_df['Ingredients'].fillna('')

# One-hot encode the 'Ingredients' column for both training and test sets
vectorizer = CountVectorizer()
X_train_vectorized = vectorizer.fit_transform(TrainSet_df['Ingredients']).toarray()
X_test_vectorized = vectorizer.transform(TestSet_df['Ingredients']).toarray()

# Convert labels to numerical values
label_encoder = LabelEncoder()
y_train_encoded = label_encoder.fit_transform(TrainSet_df['Y_encoded'])
y_test_encoded = label_encoder.transform(TestSet_df['Y_encoded'])

# Convert numerical labels to one-hot encoding
y_train_categorical = to_categorical(y_train_encoded)
y_test_categorical = to_categorical(y_test_encoded)

# Build a simple neural network model
model = Sequential()
model.add(Dense(128, activation='relu', input_dim=X_train_vectorized.shape[1]))
model.add(Dense(y_train_categorical.shape[1], activation='softmax'))

# Compile the model
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Train the model
model.fit(X_train_vectorized, y_train_categorical, epochs=5, batch_size=32)

# Evaluate the model on the test set
loss, accuracy = model.evaluate(X_test_vectorized, y_test_categorical)
print(f"Accuracy on Test Set: {accuracy * 100:.2f}%")


Epoch 1/5
72/72 [==============================] - 1s 7ms/step - loss: 0.5666 - accuracy: 0.7033
Epoch 2/5
72/72 [==============================] - 1s 7ms/step - loss: 0.4203 - accuracy: 0.8273
Epoch 3/5
72/72 [==============================] - 1s 7ms/step - loss: 0.3079 - accuracy: 0.8752
Epoch 4/5
72/72 [==============================] - 1s 7ms/step - loss: 0.2213 - accuracy: 0.9253
Epoch 5/5
31/31 [==============================] - 0s 3ms/step - loss: 0.5981 - accuracy: 0.7436
Accuracy on Test Set: 74.36%


Even better. Let us check the value predicted for the same test data.

In [17]:
test_input_index = 17
input_test_data = TestSet_df['Ingredients'].iloc[test_input_index]
actual_label = TestSet_df['Y_encoded'].iloc[test_input_index]

input_test_vectorized = vectorizer.transform([input_test_data]).toarray()

# Make a prediction on the single input
y_pred_categorical = model.predict(input_test_vectorized)

# Convert predicted labels from one-hot encoding to numerical labels
y_pred_encoded = y_pred_categorical.argmax(axis=1)

# Convert numerical labels to original labels
predicted_label = label_encoder.inverse_transform(y_pred_encoded)

# Evaluate the model on the test set
loss, accuracy = model.evaluate(X_test_vectorized, y_test_categorical)
print(f"Ingredients: {input_test_data}")
print(f"Actual Label: {'Home' if actual_label == 0 else 'Restaurant'}")
print(f"Predicted Label: {'Home' if predicted_label == 'home' else 'Restaurant'}")
print(f"Accuracy on Test Set: {accuracy * 100:.2f}%")

31/31 [==============================] - 0s 5ms/step - loss: 0.5981 - accuracy: 0.7436
Ingredients: chicken,oil,bread_crumbs,flour,egg,honey_and_garlic_sauce
Actual Label: Restaurant
Predicted Label: Restaurant
Accuracy on Test Set: 74.36%


<ipython-input-17-265c9cbb67e3>:20: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison


**Great**

We can see that the label is predicted correctly.

### Conclusion

You can train and test more models using different algorithms. We can also add new features in the feature extraction part, so that we can acheive more accuracy for both training and testing datasets.